auth0/src/main/java/com/auth0/android/dpop/DPoPProvider.kt

```diff
@@ -0,0 +1,400 @@
+        val digest = MessageDigest.getInstance("SHA-256")
+        val hash = digest.digest(input.toByteArray(Charsets.UTF_8))
+        return Base64.encodeToString(hash, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
+    }
+
+    private fun padTo32Bytes(coordinate: BigInteger): ByteArray {
+        var bytes = coordinate.toByteArray()
+        if (bytes.size > 1 && bytes[0] == 0x00.toByte()) {
+            bytes = bytes.copyOfRange(1, bytes.size)
+        }
+        if (bytes.size < 32) {
+            val paddedBytes = ByteArray(32)
+            System.arraycopy(bytes, 0, paddedBytes, 32 - bytes.size, bytes.size)
+            return paddedBytes
+        }
+        return bytes
+    }
+
+    private fun encodeBase64Url(bytes: ByteArray): String {
+        return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
+    }
+
+    private fun signData(data: ByteArray, privateKey: PrivateKey): String? {
+        try {
+            val signatureBytes = Signature.getInstance("SHA256withECDSA").run {
+                initSign(privateKey)
+                update(data)
+                sign()
+            }
+            return encodeBase64Url(convertDerToRawSignature(signatureBytes))
+        } catch (e: Exception) {
+            Log.e(TAG, "Error signing data: ${e.stackTraceToString()}")
+            throw DPoPException(DPoPException.Code.SIGNING_ERROR, e)
+        }
+    }
+
+    private fun convertDerToRawSignature(derSignature: ByteArray): ByteArray {
+        // DER format: SEQUENCE (0x30) + length + INTEGER (0x02) + length + R + INTEGER (0x02) + length + S
+        var offset = 0
+        if (derSignature[offset++] != 0x30.toByte()) throw SignatureException("Invalid DER signature: Expected SEQUENCE")
+        val length = decodeLength(derSignature, offset).also { offset += it.second }.first
+        if (length + offset != derSignature.size) throw SignatureException("Invalid DER signature: Length mismatch")
+
+        if (derSignature[offset++] != 0x02.toByte()) throw SignatureException("Invalid DER signature: Expected INTEGER for R")
+        val rLength = decodeLength(derSignature, offset).also { offset += it.second }.first
+        var r = derSignature.copyOfRange(offset, offset + rLength)
+        offset += rLength
+
+        if (derSignature[offset++] != 0x02.toByte()) throw SignatureException("Invalid DER signature: Expected INTEGER for S")
+        val sLength = decodeLength(derSignature, offset).also { offset += it.second }.first
+        var s = derSignature.copyOfRange(offset, offset + sLength)
+        offset += sLength
+
+        // Remove leading zero if present
+        if (r.size > 1 && r[0] == 0x00.toByte() && r[1].toInt() and 0x80 == 0x80) r =
+            r.copyOfRange(1, r.size)
+        if (s.size > 1 && s[0] == 0x00.toByte() && s[1].toInt() and 0x80 == 0x80) s =
+            s.copyOfRange(1, s.size)
+
+        // Pad with leading zeros to 32 bytes for P-256
```

Comment:
Is it necessary to remove the leading zero if we're going to add leading zeroes here?